React/useEffect.md

```diff
@@ -0,0 +1,186 @@
+
+  It is much more advantageous to make any API calls asychronously after the app is fully rendered on the screen.  The user gets immediate feedback that the page is loaded and data can appear as it is retrieved.
+</details>
+
+We can provide React a function to run when it is in the `componentDidMount` stage by using the `useEffect` hook as follows.
+
+```javascript
+useEffect( /* Function to run when the component is mounted */  () => {  
+    // Code to run upon mount
+  },
+    [] /* Empty Array indicates to run this once when the component is mounted. */
+);
+```
+
+### Updating
+
+When props or state changes the component is re-rendered on the screen.  We can watch for props or state variables and when they change we can use the `useEffect` hook to run a function when those variables change.  
+
+So if we wanted to run a function when our `students` state variable changes we could do:
+
+```javascript
+useEffect( /* Function to run when the `students` change */  () => {  
+    // Code to run when it changes
+  },
+    [students] /* List of variables to watch and run the callback when they change. */
+);
+```
+
+### Unmounting
+
+When a component is unmounted or removed from the DOM, sometimes you want to do operations like:
+
+1.  Save information to an external resource prior to exit
+1.  Cancel any pending API requests that haven't returned yet
+
+To provide a cleaup method for unmounting your `useEffect` callback function should return a function to call when the component is unmounted.
+
+```javascript
+useEffect( /* Function to run when the component is mounted */  () => {  
+    // Code to run upon mount
+
+    return /* Function to run on unmount */ () => {
+      // Cleanup code here
+
+    };
+  },
+    [] /* Empty Array indicates to run this once when the component is mounted. */
+);
+```
+
+## useEffect with LocalStorage
+
+Currently every time our application restarts any students we add dissapear.  This is because we are not storing the list of students in any database or other long-term storage.  We can use a feature of the browser called [localStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage).  Localstorage allows us to save key-value pairs in the browser to access when the JavaScript application restarts.
+
+### Loading Data from Local Storage
+
+We want to check localstorage for a list of students when our application starts up.  So we can use the `useEffect` hook.
+
+First we will import the `useEffect` hook.
+
+```javascript
+import React, { useState, useEffect } from 'react';
+```
+
+Instead of pulling the information from a static array we can use localStorage.  The code below will, after App is initially rendered access local storage and try to find a key `studentList`, if that key is found it converts the value into an array with `JSON.parse` and updates the state.  If the key does not exist, it uses the initial array.
```

Comment:
```suggestion
Instead of pulling the information from a static array we can use localStorage.  The code below will, after `App` is initially rendered access local storage and try to find a key `studentList`, if that key is found it converts the value into an array with `JSON.parse` and updates the state.  If the key does not exist, it uses the initial array.
```